pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -378,7 +378,12 @@ in
 lib.extendDerivation
   validity.handled
   ({
-     overrideAttrs = f: stdenv.mkDerivation (attrs // (f attrs));
+      overrideAttrs = f: let
+        f' =
+          if builtins.typeOf f == "set"
+          then (if lib.hasAttr "__functor" f then f else (_: f))
+          else f;
+      in stdenv.mkDerivation (attrs // (f' attrs));
```

Comment:
If you inline `f' attrs`, you can save some envs and closures.
According to the performance [report](https://github.com/NixOS/nixpkgs/pull/146347/checks?check_run_id=4235404519), `gc-heapSize` is up by 0.09%.
It seems fairly minor but these things add up, so I like to keep an eye out for changes to frequently called code.

Reply:
@roberth Can you make a Github suggestion that implements this change?

Reply:
Actually it won't let me, so I'll just go old school.

Starting with

```
      overrideAttrs = f: let
        f' =
          if builtins.typeOf f == "set"
          then (if lib.hasAttr "__functor" f then f else (_: f))
          else f;
      in stdenv.mkDerivation (attrs // (f' attrs));
```

then inline `f'`, saving a scope (envs in the perf report)

```
      overrideAttrs = f: stdenv.mkDerivation (attrs // ((if builtins.typeOf f == "set"
          then (if lib.hasAttr "__functor" f then f else (_: f))
          else f) attrs));
```

then apply arg `attrs`

```
      overrideAttrs = f: stdenv.mkDerivation (attrs // ((if builtins.typeOf f == "set"
          then (if lib.hasAttr "__functor" f then f attrs else (_: f) attrs)
          else f attrs)));
```

apply the constant function

```
      overrideAttrs = f: stdenv.mkDerivation (attrs // ((if builtins.typeOf f == "set"
          then (if lib.hasAttr "__functor" f then f attrs else f)
          else f attrs)));
```

remove redundant parentheses, format

```
      overrideAttrs = f: stdenv.mkDerivation (attrs // (
          if builtins.typeOf f == "set"
          then (if lib.hasAttr "__functor" f then f attrs else f)
          else f attrs));
```

avoid an attr selection `lib.hasAttr`


```
      overrideAttrs = f: stdenv.mkDerivation (attrs // (
          if builtins.typeOf f == "set"
          then (if f?__functor then f attrs else f)
          else f attrs));
```

Reply:
Oh and use builtins.isAttrs to save another application and equality

```
      overrideAttrs = f: stdenv.mkDerivation (attrs // (
          if builtins.isAttrs f
          then (if f?__functor then f attrs else f)
          else f attrs));

```

Reply:
Maybe even

```
      overrideAttrs = f: stdenv.mkDerivation (attrs // (
          if builtins.isAttrs f && !f ? __functor
          then f
          else f attrs));
```

Reply:
In that case, use de Morgan's law, to remove a _not_.

```nix
      overrideAttrs = f: stdenv.mkDerivation (attrs // (
          if builtins.isFunction f || f?__functor
          then f attrs
          else f));
```

I strongly recommend removing the whitespace around `?` just like we do with `.`. The spaces make it look like a normal binary operator, which it is not, because the rhs is an identifier, not a variable reference (or any other kind of expression).